crates/pessimistic-proof/src/certificate.rs

```diff
@@ -1,41 +1,149 @@
+        &self,
+        state: &LocalNetworkStateData,
+        signer: Address,
+    ) -> Result<MultiBatchHeader<Keccak256Hasher>, Error> {
+        let prev_balance_root = state.balance_tree.root;
+        let prev_nullifier_root = state.nullifier_set.root;
+
+        let balances_proofs: BTreeMap<TokenInfo, (U256, LocalBalancePath<Keccak256Hasher>)> = {
+            let mut res = BTreeMap::new();
+
+            let mutated_tokens = {
+                let imported_tokens = self
+                    .imported_bridge_exits
+                    .iter()
+                    .map(|exit| exit.bridge_exit.token_info);
+                let exported_tokens = self.bridge_exits.iter().map(|exit| exit.token_info);
+                imported_tokens.chain(exported_tokens)
+            };
+
+            mutated_tokens.for_each(|token| {
+                let initial_balance =
+                    U256::from_be_bytes(state.balance_tree.get(token).unwrap_or_default());
+                let path = if initial_balance.is_zero() {
+                    // TODO: dont clone once get_inclusion_proof_zero doesnt mutate anymore
+                    state
+                        .balance_tree
+                        .clone()
+                        .get_inclusion_proof_zero(token)
+                        .unwrap()
+                } else {
+                    state.balance_tree.get_inclusion_proof(token).unwrap()
+                };
+                res.insert(token, (initial_balance, path));
+            });
+            res
+        };
+
+        let imported_bridge_exits: Vec<(ImportedBridgeExit, NullifierPath<Keccak256Hasher>)> = self
+            .imported_bridge_exits
+            .iter()
+            .map(|exit| {
+                let null_key = NullifierKey {
+                    network_id: exit.global_index.network_id(),
+                    let_index: exit.global_index.leaf_index,
+                };
+                let nullifier_path = state
+                    .nullifier_set
+                    .get_non_inclusion_proof(null_key)
+                    .unwrap();
+                (exit.clone(), nullifier_path)
+            })
+            .collect();
+
+        Ok(MultiBatchHeader::<Keccak256Hasher> {
+            origin_network: self.network_id,
+            prev_local_exit_root: self.prev_local_exit_root,
+            bridge_exits: self.bridge_exits.clone(),
+            imported_bridge_exits,
+            balances_proofs,
+            prev_balance_root,
+            prev_nullifier_root,
+            signer,
+            signature: self.signature,
+            imported_rollup_exit_root: [0; 32],
+            imported_mainnet_exit_root: [0; 32],
+            imported_exits_root: None,
+            target: StateCommitment {
+                exit_root: self.new_local_exit_root,
+                balance_root: Default::default(), // computed by agglayer
+                nullifier_root: Default::default(), // computed by agglayer
+            },
```

Comment:
Not necessarily for this PR, just a general comment. I think it would be cleaner to introduce a separate type for PP input rather than recycling `MultiBatchHeader` and filling it with dummy fields. Common parts could be factored out.